modules/ingestor/src/service/advisory/csaf/mod.rs

```diff
@@ -1,2 +1,80 @@
+use super::{super::Error, Format};
+use crate::service::advisory::{csaf::loader::CsafLoader, osv::loader::OsvLoader};
+use anyhow::anyhow;
+use bytes::Bytes;
+use futures::{Stream, TryStreamExt};
+use std::time::Instant;
+use trustify_common::db::Transactional;
+use trustify_module_storage::service::{StorageBackend, SyncAdapter};
+
 pub mod loader;
 mod util;
+
+impl super::super::IngestorService {
+    pub async fn ingest<S, E>(&self, source: &str, fmt: Format, stream: S) -> Result<String, Error>
+    where
+        E: std::error::Error,
+        S: Stream<Item = Result<Bytes, E>>,
+    {
+        let start = Instant::now();
+
+        let digest = self
+            .storage
+            .store(stream)
+            .await
+            .map_err(|err| Error::Storage(anyhow!("{err}")))?;
+        let sha256 = hex::encode(digest);
+
+        let storage = SyncAdapter::new(self.storage.clone());
+        let reader = storage
+            .retrieve(sha256.clone())
+            .await
+            .map_err(Error::Storage)?
+            .ok_or_else(|| Error::Storage(anyhow!("file went missing during upload")))?;
```

Comment:
For bonus points, do we want to re-hash to also ensure file contents didn't change during the shuffling?

Reply:
I think being able to successfully retrieve it ensures that, right?

Reply:
I dunno tbh. I think it means we managed to find some file at the hash-named location. Belt-and-suspenders would then verify, I guess, if the file retrieved by hash, when hashed, hashes to the hash of the hash-based file.

Yo dawg, I heard you liked hash, so I put some hash in your hash so you can hash while you hash.

Reply:
I see what you mean, verify we got what we put.